crates/sp-domains-fraud-proof/src/verification.rs

```diff
@@ -535,20 +535,42 @@ where
     } = invalid_bundles_fraud_proof;
     let (bundle_index, is_true_invalid_fraud_proof) = (*bundle_index, *is_true_invalid_fraud_proof);
 
-    let invalid_bundle_entry = check_expected_bundle_entry::<CBlock, DomainHeader, Balance>(
+    let targeted_invalid_bundle_entry = check_expected_bundle_entry::<CBlock, DomainHeader, Balance>(
         &bad_receipt,
         bundle_index,
         invalid_bundle_type.clone(),
         is_true_invalid_fraud_proof,
     )?;
+    let bundle_extrinsic_root = targeted_invalid_bundle_entry.extrinsics_root;
+
+    // Verify the bundle proof so in following we can use the bundle dirctly
+    match proof_data {
+        InvalidBundlesProofData::Bundle(bundle_with_proof)
+        | InvalidBundlesProofData::BundleAndExecution {
+            bundle_with_proof, ..
+        } => {
+            if bundle_with_proof.bundle_index != bundle_index {
+                return Err(VerificationError::UnexpectedInvalidBundleProofData);
+            }
+            bundle_with_proof.verify::<CBlock, SKP>(domain_id, &state_root)?;
+        }
+        InvalidBundlesProofData::Extrinsic(_) => {}
+    }
+
+    // Fast path to check if the fraud proof is targetting a bad receipt that claim a non-exist extrinsic
+    // is invalid
+    if let Some(invalid_extrinsic_index) = targeted_invalid_bundle_entry.invalid_extrinsic_index() {
+        if let InvalidBundlesProofData::Bundle(bundle_with_proof) = proof_data {
+            if bundle_with_proof.bundle.extrinsics.len() as u32 <= invalid_extrinsic_index {
```

Comment:
@NingLin-P is this correct?
if `len()` is `1` then I would expect `invalid_extrinsic_index` need to be `0`, hence the comparison needs to be `<` instead of `<=` ?

Reply:
ah understood, thanks!